server/src/main/java/org/candlepin/async/tasks/EntitlerJob.java

```diff
@@ -0,0 +1,204 @@
+        this.i18n = Objects.requireNonNull(i18n);
+        this.poolCurator = Objects.requireNonNull(p);
+    }
+
+    @Override
+    public Object execute(final JobExecutionContext context) throws JobExecutionException {
+        try {
+            final JobArguments arguments = context.getJobArguments();
+            final String uuid = arguments.getAsString(CONSUMER_UUID_KEY);
+            final PoolIdAndQuantity[] poolQuantities = arguments
+                .getAs(POOL_QUANTITIES_KEY, PoolIdAndQuantity[].class);
+
+            final Map<String, Integer> poolMap = new HashMap<>();
+            for (PoolIdAndQuantity poolIdAndQuantity : poolQuantities) {
+                poolMap.put(poolIdAndQuantity.getPoolId(), poolIdAndQuantity.getQuantity());
+            }
+            final List<Entitlement> ents = entitler.bindByPoolQuantities(uuid, poolMap);
+            entitler.sendEvents(ents);
+
+            final List<PoolIdAndQuantity> consumed = ents.stream()
+                .map(ent -> new PoolIdAndQuantity(ent.getPool().getId(), ent.getQuantity()))
+                .collect(Collectors.toList());
+            poolCurator.clear();
+            return consumed;
+        }
+        catch (EntitlementRefusedException e) {
+            log.error("EntitlerJob encountered a problem, translating errors", e);
+            final Map<String, ValidationResult> validationResults = e.getResults();
+            return collectErrors(validationResults);
+        }
+        // Catch any exception that is fired and re-throw as a JobExecutionException
+        // so that the job will be properly cleaned up on failure.
+        catch (Exception e) {
+            log.error("EntitlerJob encountered a problem.", e);
+            throw new JobExecutionException(e.getMessage(), e, false);
+        }
+    }
+
+    private List<PoolIdAndErrors> collectErrors(final Map<String, ValidationResult> results) {
+        final EntitlementRulesTranslator translator = new EntitlementRulesTranslator(i18n);
+        final List<PoolIdAndErrors> poolErrors = new ArrayList<>();
+
+        for (Pool pool : poolCurator.listAllByIds(results.keySet())) {
+            final List<String> errorMessages = new ArrayList<>();
+            for (ValidationError error : results.get(pool.getId()).getErrors()) {
+                errorMessages.add(translator.poolErrorToMessage(pool, error));
+            }
+            poolErrors.add(new PoolIdAndErrors(pool.getId(), errorMessages));
+        }
+        return poolErrors;
+    }
+
+    public static EntitlerJobConfig createConfig() {
+        return new EntitlerJobConfig();
+    }
+
+    /**
+     * Job configuration object for the entitle by products job
+     */
+    public static class EntitlerJobConfig extends JobConfig {
+
+        public EntitlerJobConfig() {
+            this.setJobKey(JOB_KEY)
+                .setJobName(JOB_NAME)
+                .addConstraint(JobConstraints.uniqueByArgument(OWNER_KEY));
```

Comment:
In addition to the fact that the consumer is the target of constraining here, this job has a custom restriction in its isSchedulable/scheduleJob mechanism: 
```
        long running = jobCurator.findNumRunningByClassAndTarget(status.getTargetId(), jobClass);
        // We can start the job if there are less than N others running
        int throttle = conf.getInt(ConfigProperties.ENTITLER_JOB_THROTTLE);
        return running < throttle;
```
A new class that extends `JobConstraint` is required here, similar to the UniqueByArgConstraint (it might even extend UniqueByArgConstraint), which will not be testing that no other job of this type with the same consumer uuid is running, but that there are no more than N-1 jobs of this type of job for this consumer uuid running at the moment.

Reply:
@Januson In favor of this PR moving forward, lets remove all code relating to new job constraints, and simply change this to use a normal `.addConstraint(JobConstraints.uniqueByArgument(OWNER_KEY));`, since we need to talk about the constraints design more, and definitely work on it as a separate card.

Reply:
Done.